packages/core/src/codewhisperer/tracker/userWrittenCodeTracker.ts

```diff
@@ -0,0 +1,196 @@
+ */
+export class UserWrittenCodeTracker {
+    private _userWrittenNewCodeCharacterCount: Map<CodewhispererLanguage, number>
+    private _userWrittenNewCodeLineCount: Map<CodewhispererLanguage, number>
+    private _qIsMakingEdits: boolean
+    private _timer?: NodeJS.Timer
+    private _qUsageCount: number
+    private _lastQInvocationTime: number
+
+    static #instance: UserWrittenCodeTracker
+    static copySnippetThreshold = 50
+    static resetQIsEditingTimeoutMs = 2 * 60 * 1000
+    static defaultCheckPeriodMillis = 5 * 60 * 1000
+
+    private constructor() {
+        this._userWrittenNewCodeLineCount = new Map<CodewhispererLanguage, number>()
+        this._userWrittenNewCodeCharacterCount = new Map<CodewhispererLanguage, number>()
+        this._qUsageCount = 0
+        this._qIsMakingEdits = false
+        this._timer = undefined
+        this._lastQInvocationTime = 0
+    }
+
+    public static get instance() {
+        return (this.#instance ??= new this())
+    }
+
+    public isActive(): boolean {
+        return TelemetryHelper.instance.isTelemetryEnabled() && AuthUtil.instance.isConnected()
+    }
+
+    // this should be invoked whenever there is a successful Q feature invocation
+    // for all Q features
+    public onQFeatureInvoked() {
+        this._qUsageCount += 1
+        this._lastQInvocationTime = performance.now()
+    }
+
+    public onQStartsMakingEdits() {
+        this._qIsMakingEdits = true
+    }
+
+    public onQFinishesEdits() {
+        this._qIsMakingEdits = false
+    }
+
+    public getUserWrittenCharacters(language: CodewhispererLanguage) {
+        return this._userWrittenNewCodeCharacterCount.get(language) || 0
+    }
+
+    public getUserWrittenLines(language: CodewhispererLanguage) {
+        return this._userWrittenNewCodeLineCount.get(language) || 0
+    }
+
+    public reset() {
+        this._userWrittenNewCodeLineCount = new Map<CodewhispererLanguage, number>()
+        this._userWrittenNewCodeCharacterCount = new Map<CodewhispererLanguage, number>()
+        this._qUsageCount = 0
+        this._qIsMakingEdits = false
+        this._lastQInvocationTime = 0
+        if (this._timer !== undefined) {
+            clearTimeout(this._timer)
+            this._timer = undefined
+        }
+    }
+
+    public emitCodeContributions() {
+        const selectedCustomization = getSelectedCustomization()
+
+        for (const [language, charCount] of this._userWrittenNewCodeCharacterCount) {
+            const lineCount = this._userWrittenNewCodeLineCount.get(language) || 0
+            if (charCount > 0) {
+                client
+                    .sendTelemetryEvent({
+                        telemetryEvent: {
+                            codeCoverageEvent: {
+                                customizationArn:
+                                    selectedCustomization.arn === '' ? undefined : selectedCustomization.arn,
```

Comment:
FYI we can use `undefinedIfEmpty(s: string)` https://github.com/aws/aws-toolkit-vscode/blob/master/packages/core/src/shared/utilities/textUtilities.ts#L255-L261

Reply:
fixed